include/etcpal/cpp/thread.h

```diff
@@ -0,0 +1,435 @@
+/******************************************************************************
+ * Copyright 2019 ETC Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************
+ * This file is a part of EtcPal. For more information, go to:
+ * https://github.com/ETCLabs/EtcPal
+ ******************************************************************************/
+
+/// \file etcpal/cpp/thread.h
+/// \brief C++ wrapper and utilities for etcpal/thread.h
+
+#ifndef ETCPAL_CPP_THREAD_H_
+#define ETCPAL_CPP_THREAD_H_
+
+#include <algorithm>
+#include <cassert>
+#include <chrono>
+#include <functional>
+#include <limits>
+#include <memory>
+#include <string>
+#include <system_error>
+#include <type_traits>
+#include <utility>
+#include "etcpal/thread.h"
+#include "etcpal/cpp/common.h"
+#include "etcpal/cpp/error.h"
+
+namespace etcpal
+{
+/// \defgroup etcpal_cpp_thread thread (Threading)
+/// \ingroup etcpal_cpp
+/// \brief C++ utilities for the \ref etcpal_thread module.
+///
+/// Provides a thread class Thread which is similar to std::thread, with the key advantage that it
+/// will work on any threaded platform that EtcPal is ported for, including the embedded RTOS
+/// platforms. If your application or library does not need to run on these platforms, consider
+/// using std::thread instead.
```

Comment:
It could be good to add a link to std::jthread, a C++20 addition that has this same behavior of automatically joining in the destructor.